### ISSUE TITLE
Predicate `value is object` doesn't work in Array.prototype.filter()

### ISSUE THREAD

**Issue by Peeja** [microsoft/TypeScript#44777]
# Bug Report

### 🔎 Search Terms

type narrowing, type predicate, `value is object`, `filter()`, `lodash.isObject`

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about [type guards](https://github.com/Microsoft/TypeScript/wiki/FAQ#type-guards).

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=3.3.3#code/PTAEEsGcDkFcFsBGBTATqAhgOwCYUgPKIBWyAxgC6arKjwD2OyANsnhgGYVqgUAWtDrCyVw9LJFCIAnrwGhIGeLSxLkk8FlDNGGSHwB0AWABQpoSIpitUOEjQAKAG4ZmsZAH4AXJizSAlD4ubrRQoFgIKOgA3qagoDQUsKhaFNIADsj0HKDB7qAAvEWgAEQR9qglANymAL6m5sKi4vhEpJTOru7evgFBXaGS9CTkVLEm8YnJqRlZOXm0RQWlw+0U1XUNJiCgAKIYZHygLMjKWFRk9LDMeCi+oKujD+gY4ZFoADQK9KAA7rRkbAAcioFBoGCo4CoemOUIEqGMJkuEioLlQYlgkAAKnxNABzSTLADa0UwPgAjKBal9yV8ShhEGQSgBdGomNEY7G4rAEokABmZoAA1KByWyOVcufjIPzmQYMGzTDsAGLgZjcdE80AcejocTMWTlKKSaTgFg4SQLST-UCArQYHA4L6IWBULD0Kh45BUdKoeiZVBWdQPDiI5GQaHMZh2Y2FXIYdGSnHSgwcNUahy2d6ofxs1zR7MygXC0V5qMxtBFuUKrY7ABCru16bQ+O1uoeWANDxGlFApvNkg4GDVkgoPxoDCcKmz2r98DktGE1l4szD4gjmCjbVGhPjicxyZ5kFTzdQmcIPYoudM+e3lCrJbFN63l6r8sV2zAAE1va1L-hMEQVheB+VRUD9X5MAUfFgJOM5IXtaD4HSYDkRwKFrFcNcUXjEI4wlA9uV5AU2XAHJzzvChOhCfx-FAcZ4niBZ5U2EwgA)

### 💻 Code

```ts
// isNumber and isObject are modeled after the functions by the same names in lodash.

function isNumber(value?: any): value is number {
  return typeof value === "number";
}

function isObject(value?: any): value is object {
  return typeof value === "object";
}

// Each element could be an object or a number, so we can't treat it as either.
const variousThings = [{ a: 1 }, 1, "abc"];
variousThings[0] + 1;
variousThings[0].a;

// Filtering for only numbers yields values we can add, but not get properties of.
const allNumbers = variousThings.filter(isNumber);
allNumbers[0] + 1;
allNumbers[0].a;

// But filtering for only object yields fails to remove number from the union type.
const allObjects = variousThings.filter(isObject);
allObjects[0] + 1;
allObjects[0].a;

// Yet isObject is able to narrow a single element in a simple conditional.
const value = variousThings[0];
if (isObject(value)) {
    value.a
}
```

### 🙁 Actual behavior

`allObjects` was typed as `(string | number | { a: number })[]`, so both lines below `const allObjects` fail.

### 🙂 Expected behavior

`allObjects` should be typed as `{ a: number }[]`, so `allObjects[0].a` should pass the type checker. Notably, as seen above, `isObject` *does* manage to narrow the type of a single value with the union type to the only member of the union that's an object, it just can't narrow all of the elements using `.filter()` the way `isNumber` is able to.

I can't conceive of why they behave differently, but I also don't know the internals well. I *think* this difference is undesired, but I might be missing something deeper.

**Comment by Peeja** [comment]
Ah! I understand now. The difference is that `number` is one of the types in the union and also the type that `isNumber` checks, but the object type in the union is `{ a: number }`, while `isObject` checks for `object` in general. [If you change `isObject` to return `value is { a: number }`, it works.](https://www.typescriptlang.org/play?ts=3.3.3#code/PTAEEsGcDkFcFsBGBTATqAhgOwCYUgPKIBWyAxgC6arKjwD2OyANsnhgGYVqgUAWtDrCyVw9LJFCIAnrwGhIGeLSxLkk8FlDNGGSHwB0AWABQpoSIpitUOEjQAKAG4ZmsZAH4AXJizSAlD4ubrRQoFgIKOgA3qagoDQUsKhaFNIADsj0HKDB7qAAvEWgAEQR9qglANymAL6m5sKi4vhEpJTOru7evgFBXaGS0Zg+5VGgtaCxJvGJyakZWTl5tEUFpfQk5BTVdQ0mIKAAohhkfKAsyMpYVGT0sMx4KL6gm+1U9OgY4ZFoADQKeigADutDI2AA5FQKDQMFRwFQ9BcEQJUMYTHcJFQXKgxLBIAAVPiaADmknWAG1hhgfABGCYA2kAkoYRBkEoAXRqJhxeMJxKwZIpAAYOaAANSgWnc3n3fmkyAijkGDDc0yHABi4GY3FxgtAHE+rywzFkYzQkmk4BYOEkK0koNA4K0GBwOABiFgVCw9CoJOQVHSqHomVQVnUrw46MxkERzGYdii5NyGFxcqJCoMHG1uoctl+qH83NcCYLitFEqlxfjiYtSpVaoOYAAQl6DTm0KSDUbxKbXltKKArTbJBwMNrJBQgTQGE4VAWDcH4HJaMJrLxFtHxLHMPG2ttk7L8RnBZAsx3UHnCAOKEXTCX95Ry2LJdL73ub8+G-tDgBNAOtDe+CYIgrC8ECqioMGwKYAopJgZc1zwi6cHwOkYGYjgCLWK4W5YimISFCmabHgKQqity4A5Fej4UJ0IT+P4UxxPEBHuCqewmEAA) (Though obviously that's not a reasonable definition of `isObject`.)

I think I have a solution, though: this alternate definition of `filter` (written here as `filter2`) should work with a predicate that checks a broader type than any elements in the union, and returns the *intersection* of the two broad types (the union and the type checked by the predicate).

```ts
interface Array<T> {
  filter2<S>(
    callbackfn: (value: T | S, index: number, array: (T | S)[]) => value is S,
    thisArg?: any
  ): (T & S)[];
}
```

([Playground](https://www.typescriptlang.org/play?ts=3.3.3#code/JYOwLgpgTgZghgYwgAgIJSnAngHgCoB8yA3gLABQyyMwANpFAEw4DKBAFBVVQnLbQCNEAaxggAXMnYA3PgFcIkvMgA+yFgBpkoACYQAHpJByAtgOha4GbJPbK1LAJQBtALqPkAXiKzaC7QDO6hpc3GAAFsAB6ADmAPyScCBYoY62ygBk6i6uANwUAL4UxeQA9KWBAHKm5lDISTqBAPICAFYQCGD1UCgmAPZ6tBCNcDAMyBEoMHIgncB9IEECWBPhKAFwJiggmxBBoMi0A3AB4QB0JdOzYPMgVTXQMvIQCfXJaci+-lHIxmbQJFCPTAcigdzAWAADhA+jBPs8vJ5PMgAER-Woo-LkIrkChXOYLZptDpgJ5+F6Jd6SL4oH59YmdQGUZDA0HgqEwuE0xHIlH09qdTGFErlZAAUUQ4WQECGW3AyAQfTktEa5jeyH5JI1dTgvweUC0AT6yAA7iheCAAORdMA9OBdYBdE7Sx1rKAXciKxZdWRQeZyAJ4SIgGJBZHOYj1SQARmQBS00a0KLgAgQKLyFF9-sDwdDzgADK5kABqZDRrFZpU50B5wtnOBYiiigBidAYNeofTqC1oK3R0CCWGAMp0QRpQTNCqS9R0Oi0AjkXRAfS6MQgXUhUD60KgNz2GpgHq9ASd-Gq-ygYfhfqrQZrATONHo0EY7Ci59qjixfFoH4HBaLUtywoH8-0vAD60bMoKgAIUXag22gDsYC7DUQF7DUGS6IcRyCeA6CCMBjR6fppG2fVqC3ExVhQGZbgmDkjwWE96n4FoBTAK9KwDO8QwfJ8GFfKJ2JJL8QLYrCAgAksy2-CSOKkusGxFCoAE11yJDjAnqAQhgmY0dgwPoTXqZAAhrPSZQgOUHTuXVzJMSE9K9HRHVuPgmO9eFyS8a9s142sM3IYA4TfAIRM6MkFEcDwyGZGl62FcggA))

The biggest downside I see at the moment is that the elements come out typed as `{ a: number; } & object`. Maybe worse, the elements of `allNumbers` are typed as `number | ({ a: number; } & number)`, which is rather absurd. I'm not sure why that doesn't reduce to `number` automatically. It's perfectly functional, though; you can still treat it as a `number`.

Would a PR with this as the new definition for `.filter()`, or something similar, be welcomed?

**Comment by andrewbranch** [comment]
Your analysis is correct, and that definition for `filter` looks better in many ways. However, I’d be concerned that with the intersection, we’d be opening the doors for really explosive types (if your array element type is a union of 100 things, and your type predicate tells you that a type is a union of 100 other things, now you have a union of 100,000 things). I’m unsure whether there’s a good way forward with changing the lib typings. Semi-related is #36554, though that one actually doesn’t have any `filter` examples at the moment.

**Comment by johan** [comment]
I've had the same problem with when filtering unknown[] with type guards, which current ts just ignores. Building on [Peeja](https://github.com/Peeja)'s solution, I arrived at a solution that also addresses an edge case that broke code reliant on inferred context, changing the `src/lib/es5.d.ts` override definition:
```ts
filter<S extends T>(predicate: (value: T, index: number, array: T[]) => value is S, thisArg?: any): T[]
```
…into two overrides that both produce `(T & S)[]` instead of `T[]`:
```ts
// make array.filter(isFoo).filter(isBar) produce Array<Foo & Bar>:

interface ReadonlyArray<T> {
    filter<S extends T>(predicate: (value: T, index: number, array: readonly T[]) => value is S, thisArg?: any): (T & S)[];
    filter<S>(predicate: (value: T | S, index: number, array: readonly (T | S)[]) => value is S, thisArg?: any): (T & S)[];
}
interface Array<T> {
    filter<S extends T>(predicate: (value: T, index: number, array: T[]) => value is S, thisArg?: any): (T & S)[];
    filter<S>(predicate: (value: T | S, index: number, array: (T | S)[]) => value is S, thisArg?: any): (T & S)[];
}
```
Updated version of the PR's example, re-calibrated for current typescript 5.2.2, and with another test case that broke with [Peeja](https://github.com/Peeja)'s first pitch: [playgrounds](https://www.typescriptlang.org/play?ts=5.2.2#code/PTAEEsGcDkFcFsBGBTATqAhgOwCYUgPKIBWyAxgC6arKjwD2OyANsnhgGYVqgUAWtDrCyVw9LJFCIAnrwGhIGeLSxLkk8FlDNGGSHwB0AWABQpoSIpitUOEjQAKAG4ZmsZAH4AXJizSAlD4ubrRQoFgIKOgA3qagoDQUsKhaFNIADsj0HKDB7qAAvEWgAEQR9qglANymAL6m5sKi4vhEpJTOru7evgFBXaGS9CTkVLEm8YnJWgCEM3m0AGSLvBlZOQuFxSXD7RTVdQ0mIKAAohhkfKAsyMpYVGT0sMx4KL6gu6Mf6BjhkWgAGgU9FAAHdaGRsAByKgUGgYKjgKh6a5IgSoYwmR4SKguVBiWCQAAqfE0AHNJAVQABtaKYHwARlAtSBDKBJQwiDIJQAujVjmAAAIUSAAWmQAA9MpRxahUPR0KLRaBHs9XrRkGjAcC5GF9E8Xtc5QqvKY8QTiaSsBTqQAGHmgADUoAZ-JOwrFkulFFl8sVytVhremv42sgIP4er4BrwaD9ppM5qelvJkDtPIMGH5phOADFwMxuPjraAOAqPlhmLJylFJNJwCwcJIFpJwSrsJgcDggYhYFQsPQqGTkFR0vLMqgrOoPhxMdjIMjmMw7LXCrkMPjkyTUwYOAWiw5bP9UP5+a5l8e0-anS63UKReKpaNfeWlfhMH8KkDw7rJPq1Ua8amOeK5oFeGZZkcJx7oWaDkqW5biFWHwjJQ4T0KC4QbvKoKSBgoIYNWx6lvK8ByLQwjWKsmQJvOi7MG0oyUuum6Etu1qQLu+6OFAjGUKeOb3p6T4ynGr7KmEHafJQ34RqSf7RgBYmoAm558SK6Y3q6wFLup4GZtmAqtKhiJ4YgrC8CCqjGphvyQOSFk3HciJaHZ4DwOkFnYjgSLWK4tHiAu64hGuSZsVaNr2vy4A5IehAmZ0IT+P4oDjPE8QLJmhxmEZxCErC8hpJkQw5AASsgFwUAYhLIAAsrc9BAiiHaSkonm0NkwLKJGJb8AiUjwgA1hoqTyPZ1qsKK9BOHG4BMLkYFUWWqDKKgyFjvQ6T0JAbAJhYzRaDV9UMAAPESAB8DgcJVCrSD4DgpQU52gESQJMOkkA+PCOBIbI2DSNSDoAD6gMITB7lgbCBC9qVxOlUwpH8S4zPy9Q5XRXW0FSR0NQ4cPxCcyASMktAlMgJSgOigwQFgHBxmwJH0GRoaYEuoGoBo3DM2ssmgOzHwTgi1iSDQ8AYJorM6OCOAJuloAPYUz0gZeXGwagDgOMgj3PXMmtaclAL4zSysVJAPKmAJOUnGLg20NhhGqweUC5vQ9D+I7PGQAAQhuKUbTgsBkLQACCcqESdLsgisPuoOdCamJoRbXUHoDlRgP2VtIoeoOHF2wxM6UwUWJ0AMrXBK3C4JIF0OGObDgJC3D3QsPivTTTASj4Nbavbd0JBVGfIUSgPa8F+RhCXQKRpAodkj0-3Qw4RKgCsJf+ID-Jy0XaCl5ddc+Y3yDNwMregCDk-t5KXfHk1Yd999v3y8v5-rzyo+bBPU-ybP89+Ivy+r1fqjBO9w0DJxDnfM6z00qF24qgUu5dK5NhenvGgB8ERH3li3F6QJNAd2vl+agOc+7DzforMe1ML7Tx-j4Be90AGgDXhvI2294El1QfXQ+x8Qin3Prg3AV9PxRFvsQ+hZ9GGv3fgMd8VDv6oDnrQv+YjAHMJMGjIAA)

In other words: this problem can be addressed by [this change](https://github.com/johan/TypeScript/commit/4c0442e8244f4292e814c1c9c14b49f1bd8bb163), plus whatever test suite updates are needed.
